### PR TITLE
Synaptic Intelligence

### PR DESCRIPTION
**Pull request by karangrewal**
This PR includes

- a mixin for implementing Synaptic Intelligence (SI),
- project files that run SI in an experimental setup identical to the one reported in the [paper](http://proceedings.mlr.press/v70/zenke17a/zenke17a.pdf) (on permutedMNIST), and
- project files that run SI + dendrites on permutedMNIST.

Note that PyTorch buffers may have an obvious use case in implementing SI, but I decided not to use them. The implementation seems to be just fine without buffers, but we can revisit this option if it provides computational advantages.

### CODE REVIEW

File: projects/dendrites/permutedMNIST/experiments/centroid.py [hunk 1]
```diff
@@ -50,6 +50,11 @@ class CentroidFigure1BExperiment(CentroidFigure1B,
     pass
 
 
+class SICentroidExperiment(mixins.SynapticIntelligence,
```
Comment (subutai) [inline projects/dendrites/permutedMNIST/experiments/centroid.py:53]: Maybe move these to a separate `prototype_si.py` experiment file? We should also have an experiment file for SI only.

File: projects/dendrites/permutedMNIST/experiments/centroid.py [hunk 2]
```diff
@@ -145,10 +150,38 @@ class CentroidFigure1BExperiment(CentroidFigure1B,
     ),
 )
 
+# Synaptic Intelligence + Dendrites on 10 permutedMNIST tasks
+SI_CENTROID_10 = deepcopy(CENTROID_10)
+SI_CENTROID_10["model_args"].update(hidden_sizes=[2000, 2000])
```
Comment (subutai) [inline projects/dendrites/permutedMNIST/experiments/centroid.py:155]: We have been using hidden layers with size `[2048,2048]`

**Review by subutai — APPROVED** [review]
Looks good - left some minor comments.

File: projects/dendrites/permutedMNIST/experiments/centroid.py (hunk 2 above)
Comment (karangrewal) [reply projects/dendrites/permutedMNIST/experiments/centroid.py:155]: Both SI and SI+XdG use 2000 hidden layers per unit, that's why I changed it here, for a more fair comparison. However, our dendrite models alone (without SI) are still using 2048 hidden layers.

Comment (subutai) [reply projects/dendrites/permutedMNIST/experiments/centroid.py:155]: Ok, we may want to normalize this so we have at least some datapoints where the feed forward portions are equal. We can also include high dimensional versions as long as it's explicit.

Comment (jeremyforest) [reply projects/dendrites/permutedMNIST/experiments/centroid.py:155]: if you have SI only and SI+dendrite on the graph I think its better to have 2048 to have everything with the same number of unit and make it more comparable with what we already have. We can always explicitly specify the difference in the text and if needs be show the 2000 unit SI only in a supplementary figure.

File: projects/dendrites/permutedMNIST/experiments/centroid.py [hunk 3]
```diff
@@ -145,10 +150,38 @@ class CentroidFigure1BExperiment(CentroidFigure1B,
...
+SI_CENTROID_10.update(
+    experiment_class=SICentroidExperiment,
+    epochs=20,  # Note that SI only works well with ~20 epochs of training per task
```
Comment (jeremyforest) [inline projects/dendrites/permutedMNIST/experiments/centroid.py:158]: So here to make it comparable to what we already have should we also re-run some analysis of XdG, XdG+dendrite and dendrite only with 20 epochs to show that any difference is not also due to the number of epochs ? That would be best I think but what do you think ?

File: projects/vernon_continual/experiments/si.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+
+"""
+Experiment file that runs Synaptic Intelligence (Zenke, Poole, Ganguli (2017)) in a
```
Comment (jeremyforest) [inline projects/vernon_continual/experiments/si.py:21]: maybe adding full ref here too ?

File: projects/vernon_continual/experiments/si.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+
+def train_si_model(model, loader, optimizer, device, criterion=F.cross_entropy,
+                   post_batch_callback=None, **kwargs):
+    model.train()
+    for batch_idx, (data, target) in enumerate(loader):
+        data = data.flatten(start_dim=1)
+
+        # Share labels between tasks
+        # TODO don't hardcode num classes per task
```
Comment (jeremyforest) [inline projects/vernon_continual/experiments/si.py:51]: where is num classes hardcoded in that function ?

**Review by jeremyforest — APPROVED** [review]

File: projects/dendrites/permutedMNIST/experiments/centroid.py (hunk 2 above)
Comment (karangrewal) [reply projects/dendrites/permutedMNIST/experiments/centroid.py:155]: It seems to me that perhaps we should just re-run all experiments with 2,000 hidden units, for both a) dendrites and b) SI + dendrites. That would be a better comparison, and I doubt the hyperparameters we found would change at all. Thoughts?

File: projects/vernon_continual/experiments/si.py (hunk 5 above)
Comment (karangrewal) [reply projects/vernon_continual/experiments/si.py:51]: On the next line: `target = target % 10`. This assumes there are 10 classes per task, but not a big deal for now. I placed a 'TODO' here so that we can modify this during later code enhancements.

File: projects/dendrites/permutedMNIST/experiments/centroid.py (hunk 3 above)
Comment (karangrewal) [reply projects/dendrites/permutedMNIST/experiments/centroid.py:158]: I assume you mean SI instead of XdG, right? The number of epochs is just a hyperparameter, so I wouldn't worry too much about it. Also, they use 20 epochs in the XdG paper, too.

File: projects/vernon_continual/experiments/si.py (hunk 4 above)
Comment (karangrewal) [reply projects/vernon_continual/experiments/si.py]: Done.

File: projects/dendrites/permutedMNIST/experiments/centroid.py (hunk 1 above)
Comment (karangrewal) [reply projects/dendrites/permutedMNIST/experiments/centroid.py]: Done. Also, there is an experiment file for SI only, and that's `projects/vernon_continual/experiments/si.py`

File: nupic/research/frameworks/vernon/mixins/si.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+__all__ = ["SynapticIntelligence"]
+
+
+class SynapticIntelligence:
+    """
+    Implementation of Synaptic Intelligence (Zenke, Poole, Ganguli (2017)) that applies
+    the surrogate loss to all feed-forward parameters in a neural network.
+
+    The config should contain `si_args` should be a dict that specifies the strength of
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/vernon/mixins/si.py]: Fixed.

File: nupic/research/frameworks/vernon/mixins/si.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+import torch
+import torch.nn.functional as F
+
+__all__ = ["SynapticIntelligence"]
+
+
+class SynapticIntelligence:
+    """
+    Implementation of Synaptic Intelligence (Zenke, Poole, Ganguli (2017)) that applies
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/vernon/mixins/si.py]: Done.
